chain/src/chain.rs

```diff
@@ -280,74 +289,86 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
     ) -> Option<Fork> {
-        let mut number = new_tip_block.header().number() - 1;
+        let new_tip_number = new_tip_block.header().number();
         let mut old_blocks = Vec::new();
         let mut new_blocks = Vec::new();
         let mut open_exts = Vec::new();
 
         if new_tip_ext.valid.is_none() {
             open_exts.push(new_tip_ext);
         } else {
-            // must be Some(false)
+            // txs in block are invalid
             return None;
         }
 
-        let mut is_open = true;
+        let mut is_new_record = true;
 
-        // The old fork may longer than new fork
-        if number < current_tip_number {
-            for n in number..=current_tip_number {
-                let hash = self.shared.block_hash(n).unwrap();
+        // new_blocks = forks[latest_common + 1 .. new_tip]
+        // old_blocks = chain[latest_common + 1 .. old_tip]
+        new_blocks.push(new_tip_block.clone());
+        let mut index = new_tip_number - 1;
+        let mut index_hash = new_tip_block.header().parent_hash().clone();
+
+        // if new_tip_number <= current_tip_number
+        // then old_blocks.extend(chain[new_tip_number .. =current_tip_number])
+        // if new_tip_number > current_tip_number
+        // then new_blocks.extend(forks[current_tip_number + 1 .. =new_tip_number])
+        if new_tip_number <= current_tip_number {
+            for bn in new_tip_number..=current_tip_number {
+                let hash = self.shared.block_hash(bn).unwrap();
                 let old_block = self.shared.block(&hash).unwrap();
-
                 old_blocks.push(old_block);
             }
-        }
-
-        //TODO: remove this clone
-        new_blocks.push(new_tip_block.clone());
+        } else {
+            while index > current_tip_number {
+                let new_block = self.shared.block(&index_hash).unwrap();
+
+                if is_new_record {
+                    let ext = self.shared.block_ext(&index_hash).unwrap();
+                    if ext.valid.is_none() {
+                        open_exts.push(ext)
+                    } else {
+                        is_new_record = false;
+                    }
+                }
 
-        let mut hash = new_tip_block.header().parent_hash().clone();
+                index_hash = new_block.header().parent_hash().clone();
+                new_blocks.push(new_block);
+                index -= 1;
+            }
+        }
 
+        // find latest common ancestor
         loop {
-            if number <= current_tip_number {
-                let old_hash = self.shared.block_hash(number).unwrap();
-
-                if old_hash == hash {
-                    break;
-                }
-
-                let old_block = self.shared.block(&old_hash).unwrap();
-                old_blocks.push(old_block);
+            if index == 0 {
+                break;
+            }
+            let old_hash = self.shared.block_hash(index).unwrap();
```

Comment:
`common_parent_hash` or something like it is a more related name.

Reply:
It's NOT `common_parent_hash`